File: problem/Best_Time/maximumProfit.py
class Solution:
    def maximumProfit(self, A, k):
        res = [0] * (k + 1)
        bought = [-10**9] * k
        sold = [0] * k

        for a in A:
            for j in range(k, 0, -1):
                res[j] = max(res[j], bought[j - 1] + a, sold[j - 1] - a)
                bought[j - 1] = max(bought[j - 1], res[j - 1] - a)
                sold[j - 1] = max(sold[j - 1], res[j - 1] + a)
        
        return max(res)

File: problem/Best_Time/maximumProfit.java
class Solution {
    public long maximumProfit(int[] A, int k) {
        long[] res = new long[k + 1], bought = new long[k], sold = new long[k];
        Arrays.fill(bought, -1000000000);
        for (int a : A) {
            for (int j = k; j >= 1; j--) {
                res[j] = Math.max(res[j], Math.max(bought[j - 1] + a, sold[j - 1] - a));
                bought[j - 1] = Math.max(bought[j - 1], res[j - 1] - a);
                sold[j - 1] = Math.max(sold[j - 1], res[j - 1] + a);
            }
        }

        long ans = 0;
        for (long v : res)
            ans = Math.max(ans, v);
        return ans;
    }
}